### PR TITLE
[no sq] Bulk LBMs

### PR DESCRIPTION
**Pull request by sfan5**
previously: #12128
related: #14723

## To do

This PR is a Ready for Review :rocket: 

## How to test

1.
```lua
minetest.register_lbm({
	name = ":lol",
	nodenames = {"default:apple"},
	run_at_every_load = true,
	bulk_action = function(pos_list, dtime_s)
		print("LBM on " .. #pos_list .. " nodes")
		core.bulk_set_node(pos_list, {name="default:cloud"})
	end,
})
```

2. check that non-bulk LBMs still work

### CODE REVIEW

**Review by SmallJoker — COMMENTED** [review]
Tested with [script N°172](https://github.com/SmallJoker/minetest_lab/blob/master/init_172_bulk_lbms_14954.lua)

Works as intended.

lbms

File: doc/lua_api.md [hunk 1]
```diff
@@ -9130,7 +9132,12 @@ gets activated (not loaded!)
     action = function(pos, node, dtime_s) end,
     -- Function triggered for each qualifying node.
     -- `dtime_s` is the in-game time (in seconds) elapsed since the block
-    -- was last active
+    -- was last active.
+
+    bulk_action = function(pos_list, dtime_s) end,
+    -- Function triggered with a list of all applicable node positions at once.
+    -- This can be set alternatively to `action` (not both). Available since 5.10.0.
```
[the remark this answers is not in the document]
Comment (sfan5) [reply doc/lua_api.md:9139]: I think the first sentence is clear as-is, unlike saying it "overrides".

File: src/serverenvironment.cpp [hunk 2]
```diff
@@ -255,41 +253,73 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block,
...
-	content_t c;
-	auto it = getLBMsIntroducedAfter(stamp);
-	for (; it != m_lbm_lookup.end(); ++it) {
-		// Cache previous version to speedup lookup which has a very high performance
-		// penalty on each call
+
+	// Collect a list of all LBMs and associated positions
+	struct LBMToRun {
+		std::unordered_set<v3s16> p;
+		std::unordered_set<LoadingBlockModifierDef*> l;
```
[the remark this answers is not in the document]
Comment (sfan5) [reply src/serverenvironment.cpp]: added a comment instead

File: doc/lua_api.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (SmallJoker) [reply doc/lua_api.md]: The first sentence does not clarify whether both are executed or which one takes priority.

[the remark this answers is not in the document]
Comment (SmallJoker) [reply doc/lua_api.md]: Well then ... let's hope it's clear enough :shrug:

Thanks for the changes.

**Review by SmallJoker — APPROVED** [review]

File: src/script/cpp_api/s_env.cpp [hunk 3]
```diff
@@ -177,21 +280,9 @@ void ScriptApiEnv::initializeEnvironment(ServerEnvironment *env)
 		int id = lua_tonumber(L, -2);
 		int current_lbm = lua_gettop(L);
 
-		std::set<std::string> trigger_contents;
+		std::vector<std::string> trigger_contents;
```
Comment (Desour) [inline src/script/cpp_api/s_env.cpp:284]: Looks like new code is not removing duplicates.

File: doc/lua_api.md [hunk 4]
```diff
@@ -9148,7 +9150,13 @@ gets activated (not loaded!)
...
     -- Function triggered for each qualifying node.
     -- `dtime_s` is the in-game time (in seconds) elapsed since the block
-    -- was last active
+    -- was last active.
+
+    bulk_action = function(pos_list, dtime_s) end,
+    -- Function triggered with a list of all applicable node positions at once.
+    -- This can be provided as an alternative to `action` (not both).
+    -- Available since `minetest.features.bulk_lbms` (5.10.0)
+    -- `dtime_s`: see above
```
Comment (Desour) [inline doc/lua_api.md:9159]: If one lbm adds more nodes, are these nodes included in the remaining lbms of the current block activation?
(Edit: Looks like no.)

I think we should specify such things, or explicitly write down that it's unspecified.

Also:
* Are positions guaranteed to be in one block?
* If in one block, is it guaranteed that it's all applicable nodes in the block?
* Are there order guarantees?
  (Please guarantee none of the above.)
* Btw, for future optimization, a flag in definition to allow false positives might be cool.

File: src/serverenvironment.cpp [hunk 5]
```diff
@@ -254,40 +254,72 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block,
...
 			if (!lbm_list)
 				continue;
-			for (auto lbmdef : *lbm_list) {
-				lbmdef->trigger(env, pos + pos_of_block, n, dtime_s);
-				if (block->isOrphan())
-					return;
-				n = block->getNodeNoCheck(pos);
-				if (n.getContent() != c)
-					break; // The node was changed and the LBMs no longer apply
+			auto &batch = to_run[c];
```
Comment (Desour) [inline src/serverenvironment.cpp:287]: No hashtable lookup needed if c hasn't changed. Should be cached.

File: src/serverenvironment.cpp [hunk 6]
```diff
@@ -254,40 +254,72 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block,
...
+			batch.p.insert(pos);
+			if (c_changed) {
+				batch.l.insert(lbm_list->begin(), lbm_list->end());
```
Comment (Desour) [inline src/serverenvironment.cpp:290]: Insert only needed for new batches (i.e. when `to_run.insert()` returns [it,true]).

File: src/serverenvironment.cpp [hunk 7]
```diff
@@ -254,40 +254,72 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block,
...
 
-	const v3s16 pos_of_block = block->getPosRelative();
-	v3s16 pos;
-	MapNode n;
-	content_t c;
-	auto it = getLBMsIntroducedAfter(stamp);
+	// Collect a list of all LBMs and associated positions
+	struct LBMToRun {
+		std::unordered_set<v3s16> p; // node positions
+		std::unordered_set<LoadingBlockModifierDef*> l;
```
Comment (Desour) [inline src/serverenvironment.cpp:260]: Why not vector?

File: src/serverenvironment.h [hunk 8]
```diff
@@ -100,19 +100,25 @@ struct LoadingBlockModifierDef
...
-	void deleteContents();
+	LBMContentMapping() = default;
 	void addLBM(LoadingBlockModifierDef *lbm_def, IGameDef *gamedef);
 	const lbm_map::mapped_type *lookup(content_t c) const;
+	const lbm_vector &getList() const { return lbm_list; }
+
+	// This struct owns the LBM pointers.
+	~LBMContentMapping();
+	DISABLE_CLASS_COPY(LBMContentMapping);
+	ALLOW_CLASS_MOVE(LBMContentMapping);
```
[the remark this answers is not in the document]
Comment (red-001) [reply src/serverenvironment.h:122]: what's wrong with it? Should just leave an empty map and vector in the moved-from class.

[the remark this answers is not in the document]
Comment (red-001) [reply src/serverenvironment.h:122]: Checking `cppreference` it seems the standard only guarantees the vector will be empty after the move, I suppose the map could in theory have any data in it even completely invalid data as long as it was valid itself. Lucky the destructor only uses the vector so this still seems standard compliant.

[the remark this answers is not in the document]
Comment (Desour) [reply src/serverenvironment.h:122]: Right. I was stupid, I somehow forgot that the members are the containers and not the pointers.

File: src/script/cpp_api/s_env.cpp (hunk 3 above)
Comment (sfan5) [reply src/script/cpp_api/s_env.cpp:284]: correct, does it matter?

File: src/serverenvironment.cpp (hunk 7 above)
Comment (sfan5) [reply src/serverenvironment.cpp:260]: with the current structure insert is potentially called multiple times

File: src/serverenvironment.cpp (hunk 6 above)
Comment (sfan5) [reply src/serverenvironment.cpp:290]: this assumption does not hold due to the outer loop

**Review by Desour — APPROVED** [review]
Looks good. :+1:

**Pull request merged**